### PR TITLE
Fix race condition between reprocessing and pausing stream processor

### LINKED ISSUE #5149: NPE in JobTimeoutTrigger.onResumed
**Describe the bug**

It seems that it happen quite often that the StreamProcessor fails because of an NPE in the JobTimeoutTrigger.
My current assumption is that this happens after the Broker went into reject commands mode. After disk usage becomes lower again then the StreamProcessor is resumed, but this seem to fail sometimes in the JobTimeoutTrigger.

```
 Loading... Scanned up to 2019-07-27 (09:59:59.999)  
2020-07-20 03:47:17.613 CEST
Actor Broker-2-StreamProcessor-1 failed in phase STARTED.
2020-07-25 17:51:14.897 CEST
Actor Broker-2-StreamProcessor-1 failed in phase STARTED.
2020-07-25 17:55:40.150 CEST
Actor Broker-2-StreamProcessor-3 failed in phase STARTED.
2020-07-25 20:07:12.466 CEST
Actor Broker-1-StreamProcessor-1 failed in phase STARTED.
2020-07-25 21:05:13.824 CEST
Actor Broker-1-StreamProcessor-1 failed in phase STARTED.
2020-07-26 02:58:09.414 CEST
Actor Broker-1-StreamProcessor-1 failed in phase STARTED.
2020-07-26 12:24:10.074 CEST
Actor Broker-0-StreamProcessor-1 failed in phase STARTED.
2020-08-02 01:15:12.378 CEST
Actor Broker-2-StreamProcessor-2 failed in phase STARTED.
2020-08-03 01:31:45.105 CEST
Actor Broker-2-StreamProcessor-2 failed in phase STARTED.
2020-08-04 02:28:46.666 CEST
Actor Broker-0-StreamProcessor-2 failed in phase STARTED.
2020-08-06 08:04:03.299 CEST
Actor Broker-1-StreamProcessor-2 failed in phase STARTED.
2020-08-06 22:39:56.653 CEST
Actor Broker-2-StreamProcessor-2 failed in phase STARTED.
2020-08-07 04:47:06.717 CEST
Actor Broker-1-StreamProcessor-1 failed in phase STARTED.
2020-08-08 21:51:57.202 CEST
Actor Broker-2-StreamProcessor-2 failed in phase STARTED.
2020-08-10 11:31:01.538 CEST
Actor Broker-1-StreamProcessor-2 failed in phase STARTED.
2020-08-11 07:10:04.669 CEST
Actor Broker-0-StreamProcessor-2 failed in phase STARTED.
```

See https://console.cloud.google.com/errors/CLXBrJHsz7fFfg?time=P1D&project=zeebe-io&authuser=1

**To Reproduce**
Run a benchmark and reach the disk limit.

**Expected behavior**
Resuming the StreamProcessor without issues.

**Log/Stacktrace**

<details><summary>Full Stacktrace</summary>
 <p>

```
java.lang.NullPointerException: null
at io.zeebe.engine.processing.job.JobTimeoutTrigger.onResumed (JobTimeoutTrigger.java:63)
at java.util.ArrayList.forEach (Unknown Source)
at io.zeebe.engine.processing.streamprocessor.StreamProcessor.lambda$resumeProcessing$4 (StreamProcessor.java:354)
at io.zeebe.util.sched.ActorControl.lambda$call$0 (ActorControl.java:151)
at io.zeebe.util.sched.ActorJob.invoke (ActorJob.java:62)
at io.zeebe.util.sched.ActorJob.execute (ActorJob.java:39)
at io.zeebe.util.sched.ActorTask.execute (ActorTask.java:118)
at io.zeebe.util.sched.ActorThread.executeCurrentTask (ActorThread.java:107)
at io.zeebe.util.sched.ActorThread.doWork (ActorThread.java:91)
at io.zeebe.util.sched.ActorThread.run (ActorThread.java:204)
```

</p>
</details>

**Environment:**
 - OS: k8s
 - Zeebe Version: snapshot
 - Configuration: default benchmark

### LINKED ISSUE #5156: Race condition between pausing stream processor and reprocessing
**Describe the bug**

During investigating https://github.com/zeebe-io/zeebe/issues/5149 we saw that the StreamProcessor can during reprocessing be paused. This will not interrupt reprocessing, which means after reprocessing ends it will change the state of the processor again to PROCESSING.

_Related code:_

Pausing
https://github.com/zeebe-io/zeebe/blob/develop/engine/src/main/java/io/zeebe/engine/processing/streamprocessor/StreamProcessor.java#L345
After reprocessing
 https://github.com/zeebe-io/zeebe/blob/develop/engine/src/main/java/io/zeebe/engine/processing/streamprocessor/StreamProcessor.java#L128

**To Reproduce**
Have broker with a large log and restart the broker, such that reprocessing takes a while.

**Expected behavior**
After reprocessing nothing happens, since processor is paused.

**Environment:**
 - OS: [e.g. Linux]
 - Zeebe Version: snapshot
 - Configuration: [e.g. exporters etc.]

### PR DESCRIPTION
**Pull request by ChrisKujawa**
## Description

Add new reprocessing tests to reproduce two bugs, which happen during reprocessing and pausing/resuming the stream processor. Changed default behavior of stopping stream processor, after closing stream processor runtime is deleted.
Deletion of runtime actually causes reprocessing, before it hasn't happened.
Add new method to take snapshot explicitly in test, to test recovery with snapshot.

The bug issue happened during reprocessing, when the stream processor was paused. It happen that after reprocessing the state was reset to processing, instead of paused. If resuming before reprocessing was done, this caused NPE's, since processors/listeners haven't been initialized yet.

Changed stream processor life cycle aware guarantees, it is now guaranteed that `onRecovered` is called as first method. After that `onPaused` can happen and `onResume`. Before we haven't had these guarantees.

## Related issues

closes https://github.com/zeebe-io/zeebe/issues/5156
closes https://github.com/zeebe-io/zeebe/issues/5149

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [X] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to backport the fix to the last two minor versions

Testing:
* [X] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the release announcement

### CODE REVIEW

File: engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorReprocessingTest.java [hunk 1]
```diff
@@ -87,6 +92,85 @@ public void shouldCallRecordProcessorLifecycle() {
...
+        streamProcessorRule.startTypedStreamProcessor(
+            (processors, context) ->
+                processors.withListener(
+                    new StreamProcessorLifecycleAware() {
+                      @Override
+                      public void onRecovered(final ReadonlyProcessingContext context) {
+                        countDownLatch.countDown();
+                      }
+                    }));
+    streamProcessor.pauseProcessing();
```
Comment (saig0) [inline engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorReprocessingTest.java:125]: Please add some visual separation. I oversaw the statement :sweat_smile:

File: engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorReprocessingTest.java [hunk 2]
```diff
@@ -87,6 +92,85 @@ public void shouldCallRecordProcessorLifecycle() {
...
+    final var success = countDownLatch.await(15, TimeUnit.SECONDS);
+
+    // then
+    assertThat(success).isTrue();
+    final var currentPhase = streamProcessor.getCurrentPhase().join();
+    assertThat(currentPhase).isEqualTo(Phase.PAUSED);
```
Comment (saig0) [inline engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorReprocessingTest.java:131]: I don't like to verify the behavior by exposing an internal state. Maybe, the `currentPhase` is set correctly but the processor behaves wrong. 

Can we test the behavior instead? Like in the other test, write a record and verify that it is not read by the stream processor.

File: engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorReprocessingTest.java [hunk 3]
```diff
@@ -87,6 +92,85 @@ public void shouldCallRecordProcessorLifecycle() {
...
+                      }
+                    }));
+    streamProcessor.pauseProcessing();
+    streamProcessor.resumeProcessing();
+    final var success = countDownLatch.await(15, TimeUnit.SECONDS);
+
+    // then
+    assertThat(success).isTrue();
+    final var currentPhase = streamProcessor.getCurrentPhase().join();
+    assertThat(currentPhase).isEqualTo(Phase.PROCESSING);
```
Comment (saig0) [inline engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorReprocessingTest.java:171]: Can we add a test to verify that the `StreamProcessorLifecycleAware` is not invoked before reprocessing is done?

**Review by saig0 — CHANGES_REQUESTED** [review]
Nice solution :+1: 

Just some comments on the tests :)

File: engine/src/test/java/io/zeebe/engine/processing/streamprocessor/EngineReprocessingTest.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                    .withVariable("data", RandomString.make(1024))
+                    .create());
+
+    Awaitility.await()
+        .until(
+            () ->
+                RecordingExporter.workflowInstanceRecords()
+                    .withElementType(BpmnElementType.PROCESS)
+                    .withIntent(ELEMENT_ACTIVATED)
+                    .count(),
```
[the remark this answers is not in the document]
Comment (Zelldon) [reply engine/src/test/java/io/zeebe/engine/processing/streamprocessor/EngineReprocessingTest.java:53]: ah thanks

File: engine/src/test/java/io/zeebe/engine/processing/streamprocessor/EngineReprocessingTest.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+  @Before
+  public void init() {
+    engineRule.deployment().withXmlResource(SIMPLE_FLOW).deploy();
+    IntStream.range(0, 100)
+        .forEach(
+            i ->
+                engineRule
+                    .workflowInstance()
+                    .ofBpmnProcessId(PROCESS_ID)
+                    .withVariable("data", RandomString.make(1024))
```
[the remark this answers is not in the document]
Comment (Zelldon) [reply engine/src/test/java/io/zeebe/engine/processing/streamprocessor/EngineReprocessingTest.java:43]: Tbh I just played a bit around to make the reprocessing take a bit longer. What is your concern? Actually we can build it once.

File: engine/src/test/java/io/zeebe/engine/processing/streamprocessor/EngineReprocessingTest.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+    Awaitility.await("Await reprocessing of " + lastSize)
+        .until(() -> RecordingExporter.getRecords().size(), (size) -> size >= lastSize);
+
+    // when - then
+    engineRule.workflowInstance().ofBpmnProcessId(PROCESS_ID).create();
+  }
+
+  @Test
+  public void shouldNotContinueProcessingWhenPausedDuringReprocessing() {
```
[the remark this answers is not in the document]
Comment (Zelldon) [reply engine/src/test/java/io/zeebe/engine/processing/streamprocessor/EngineReprocessingTest.java:88]: How?

File: engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorReprocessingTest.java (hunk 3 above)
Comment (Zelldon) [reply engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorReprocessingTest.java:171]: Then we have the same issue as above right long timeout? :sweat_smile:

File: engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorReprocessingTest.java (hunk 2 above)
Comment (Zelldon) [reply engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorReprocessingTest.java:131]: Then we have the same issue as above right long timeout? :sweat_smile:

File: engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorReprocessingTest.java (hunk 3 above)
Comment (Zelldon) [reply engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorReprocessingTest.java:171]: Via Mockito? Or how you would do that?

File: engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorReprocessingTest.java (hunk 1 above)
Comment (Zelldon) [reply engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorReprocessingTest.java:125]: valid point :D

**Comment by Zelldon** [comment]
@saig0 thanks for the review lets discuss the review via zoom

File: engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorReprocessingTest.java (hunk 3 above)
Comment (saig0) [reply engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorReprocessingTest.java:171]: Yes. Or, using a simple list and add the invocations in the list.

File: engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorReprocessingTest.java (hunk 2 above)
Comment (saig0) [reply engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorReprocessingTest.java:131]: :see_no_evil:

File: engine/src/test/java/io/zeebe/engine/processing/streamprocessor/EngineReprocessingTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (saig0) [reply engine/src/test/java/io/zeebe/engine/processing/streamprocessor/EngineReprocessingTest.java:88]: I don't know :see_no_evil:

**Comment by Zelldon** [comment]
Would be cool if we can take a look again tomorrow

**Comment by Zelldon** [comment]
We discussed the tests and decided to verify the behavior instead of the state, which means we wait until timeout happens so we make sure that no longer processing happens. It is not ideal but as we know testing something which should not happen is not ideal, but here we can be kind of sure that after some seconds nothing more will happen. Alternatively we could verify the state (the flag) but this doesn't give guarantees that we break the implementation and introduce a bug where we then have again a different behavior but still the state.

**Review by saig0 — APPROVED** [review]
Good job :+1: 

Just one minor optional comment.

**Pull request merged**